### ISSUE TITLE
no output results

### ISSUE THREAD

**Issue by kaiseu** [jvm-profiling-tools/async-profiler#231]
there is no output results after running below command on RHEL7.6 but when using start and then stop there's output:

./profiler.sh -e cpu -d 30 -f test1.svg 72921

[root@aep-slave1 async-profiler]# java -agentpath:./build/libasyncProfiler.so=start,summary,flat -version
Started [cpu] profiling
java version "1.8.0_181"
Java(TM) SE Runtime Environment (build 1.8.0_181-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.181-b13, mixed mode)
--- Execution profile ---
Total samples       : 0

Frame buffer usage  : 0%

          ns  percent  samples  top
  ----------  -------  -------  ---

[root@aep-slave1 async-profiler]# ./profiler.sh start 72921
Started [cpu] profiling
[root@aep-slave1 async-profiler]# ./profiler.sh stop 72921
--- Execution profile ---
Total samples       : 2

Frame buffer usage  : 0.0011%

--- 10002191 ns (50.04%), 1 sample
  [ 0] G1CollectorPolicy::predict_bytes_to_copy(HeapRegion*)
  [ 1] G1CollectorPolicy::predict_region_elapsed_time_ms(HeapRegion*, bool)
  [ 2] G1CollectorPolicy::update_incremental_cset_info(HeapRegion*, unsigned long)
  [ 3] YoungList::rs_length_sampling_next()
  [ 4] ConcurrentG1RefineThread::sample_young_list_rs_lengths()
  [ 5] ConcurrentG1RefineThread::run_young_rs_sampling()
  [ 6] ConcurrentG1RefineThread::run()
  [ 7] java_start(Thread*)
  [ 8] start_thread

--- 9984714 ns (49.96%), 1 sample
  [ 0] PerfEvents::destroyForThread(int)
  [ 1] [unknown]

          ns  percent  samples  top
  ----------  -------  -------  ---
    10002191   50.04%        1  G1CollectorPolicy::predict_bytes_to_copy(HeapRegion*)
     9984714   49.96%        1  PerfEvents::destroyForThread(int)

already set below configuration:
[root@aep-slave1 async-profiler]# cat /proc/sys/kernel/perf_event_paranoid
1
[root@aep-slave1 async-profiler]# cat /proc/sys/kernel/kptr_restrict
0

**Comment by apangin** [comment]
I presume you're trying to measure an idle application (that doesn't consume CPU), hence no samples in `cpu` profiling mode. Wall clock mode `-e wall` will sample idle threads, too.

The default profiling interval is 10ms CPU time. That's not enough to collect even a single sample while running `java -version`. Try reducing interval, e.g.
```
java -agentpath:./build/libasyncProfiler.so=start,summary,flat,traces,interval=100us -version
```
and you'll likely see some samples:
```
Started [cpu] profiling
java version "1.8.0_211"
Java(TM) SE Runtime Environment (build 1.8.0_211-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.211-b12, mixed mode)
--- Execution profile ---
Total samples       : 18

Frame buffer usage  : 0.0209%

--- 1040401 ns (23.04%), 1 sample
  [ 0] JvmtiEnv::GetClassMethods(oopDesc*, int*, _jmethodID***)
  [ 1] jvmti_GetClassMethods
  [ 2] VM::loadMethodIDs(_jvmtiEnv*, _jclass*)
  [ 3] InstanceKlass::link_class_impl(instanceKlassHandle, bool, Thread*)
  [ 4] InstanceKlass::initialize_impl(instanceKlassHandle, Thread*)
  [ 5] InstanceKlass::initialize(Thread*)
  [ 6] LinkResolver::resolve_field(fieldDescriptor&, KlassHandle, Symbol*, Symbol*, KlassHandle, Bytecodes::Code, bool, bool, Thread*)
  [ 7] LinkResolver::resolve_field_access(fieldDescriptor&, constantPoolHandle, int, Bytecodes::Code, Thread*)
  [ 8] InterpreterRuntime::resolve_get_put(JavaThread*, Bytecodes::Code)
  [ 9] java.lang.invoke.MethodHandleImpl.initStatics
  [10] java.lang.invoke.MethodHandle.<clinit>

--- 843526 ns (18.68%), 1 sample
  [ 0] _raw_spin_unlock_irqrestore_[k]
  [ 1] try_to_wake_up_[k]
  [ 2] wake_up_q_[k]
  [ 3] do_futex_[k]
  [ 4] __x64_sys_futex_[k]
  [ 5] do_syscall_64_[k]
  [ 6] entry_SYSCALL_64_after_hwframe_[k]
  [ 7] __pthread_cond_signal
  [ 8] CompileBroker::compile_method_base(methodHandle, int, int, methodHandle, int, char const*, Thread*)
  [ 9] CompileBroker::compile_method(methodHandle, int, int, methodHandle, int, char const*, Thread*)
  [10] AdvancedThresholdPolicy::submit_compile(methodHandle, int, CompLevel, JavaThread*)
  [11] SimpleThresholdPolicy::compile(methodHandle, int, CompLevel, JavaThread*) [clone .part.24]
  [12] AdvancedThresholdPolicy::method_back_branch_event(methodHandle, methodHandle, int, CompLevel, nmethod*, JavaThread*)
  [13] SimpleThresholdPolicy::event(methodHandle, methodHandle, int, int, CompLevel, nmethod*, JavaThread*)
  [14] InterpreterRuntime::frequency_counter_overflow_inner(JavaThread*, unsigned char*)
  [15] InterpreterRuntime::frequency_counter_overflow(JavaThread*, unsigned char*)
  [16] java.lang.String.equals
  [17] java.util.Hashtable.get
  [18] java.util.Properties.getProperty
  [19] java.lang.System.getProperty

--- 770032 ns (17.05%), 3 samples
  [ 0] _raw_spin_unlock_irqrestore_[k]
  [ 1] wake_up_new_task_[k]
  [ 2] _do_fork_[k]
  [ 3] __x64_sys_clone_[k]
  [ 4] do_syscall_64_[k]
  [ 5] entry_SYSCALL_64_after_hwframe_[k]
  [ 6] clone
  [ 7] [unknown]

--- 460051 ns (10.19%), 1 sample
  [ 0] ClassFileParser::parse_constant_pool(Thread*)
  [ 1] ClassFileParser::parseClassFile(Symbol*, ClassLoaderData*, Handle, KlassHandle, GrowableArray<Handle>*, TempNewSymbol&, bool, Thread*)
  [ 2] ClassLoader::load_classfile(Symbol*, Thread*)
  [ 3] SystemDictionary::load_instance_class(Symbol*, Handle, Thread*)
  [ 4] SystemDictionary::resolve_instance_class_or_null(Symbol*, Handle, Handle, Thread*)
  [ 5] SystemDictionary::resolve_or_fail(Symbol*, Handle, Handle, bool, Thread*)
  [ 6] ConstantPool::klass_at_impl(constantPoolHandle, int, Thread*)
  [ 7] ConstantPool::klass_ref_at(int, Thread*)
  [ 8] LinkResolver::resolve_invokestatic(CallInfo&, constantPoolHandle, int, Thread*)
  [ 9] LinkResolver::resolve_invoke(CallInfo&, Handle, constantPoolHandle, int, Bytecodes::Code, Thread*)
  [10] InterpreterRuntime::resolve_invoke(JavaThread*, Bytecodes::Code)
  [11] java.lang.invoke.MethodHandle.<clinit>

--- 267941 ns (5.93%), 2 samples
  [ 0] native_write_msr_[k]
  [ 1] x2apic_send_IPI_[k]
  [ 2] native_send_call_func_single_ipi_[k]
  [ 3] generic_exec_single_[k]
  [ 4] smp_call_function_single_[k]
  [ 5] task_function_call_[k]
  [ 6] perf_install_in_context_[k]
  [ 7] __do_sys_perf_event_open_[k]
  [ 8] __x64_sys_perf_event_open_[k]
  [ 9] do_syscall_64_[k]
  [10] entry_SYSCALL_64_after_hwframe_[k]
  [11] syscall

--- 201627 ns (4.46%), 1 sample
  [ 0] _raw_spin_unlock_irqrestore_[k]
  [ 1] pty_write_[k]
  [ 2] n_tty_write_[k]
  [ 3] tty_write_[k]
  [ 4] __vfs_write_[k]
  [ 5] vfs_write_[k]
  [ 6] ksys_write_[k]
  [ 7] __x64_sys_write_[k]
  [ 8] do_syscall_64_[k]
  [ 9] entry_SYSCALL_64_after_hwframe_[k]
  [10] write
  [11] writeBytes
  [12] Java_java_io_FileOutputStream_writeBytes
  [13] java.io.FileOutputStream.writeBytes
  [14] java.io.FileOutputStream.write
  [15] java.io.BufferedOutputStream.flushBuffer
  [16] java.io.BufferedOutputStream.flush
  [17] java.io.PrintStream.write
  [18] sun.nio.cs.StreamEncoder.writeBytes
  [19] sun.nio.cs.StreamEncoder.implFlushBuffer
  [20] sun.nio.cs.StreamEncoder.flushBuffer
  [21] java.io.OutputStreamWriter.flushBuffer
  [22] java.io.PrintStream.write
  [23] java.io.PrintStream.print
  [24] java.io.PrintStream.println
  [25] sun.misc.Version.print
  [26] sun.misc.Version.print

--- 172544 ns (3.82%), 1 sample
  [ 0] _raw_spin_unlock_irqrestore_[k]
  [ 1] try_to_wake_up_[k]
  [ 2] wake_up_q_[k]
  [ 3] do_futex_[k]
  [ 4] __x64_sys_futex_[k]
  [ 5] do_syscall_64_[k]
  [ 6] entry_SYSCALL_64_after_hwframe_[k]
  [ 7] __pthread_cond_signal
  [ 8] WatcherThread::stop()
  [ 9] before_exit(JavaThread*)
  [10] Threads::destroy_vm()
  [11] jni_DestroyJavaVM
  [12] JavaMain
  [13] start_thread

--- 158566 ns (3.51%), 2 samples
  [ 0] smp_call_function_single_[k]
  [ 1] task_function_call_[k]
  [ 2] event_function_call_[k]
  [ 3] _perf_event_enable_[k]
  [ 4] _perf_event_refresh_[k]
  [ 5] _perf_ioctl_[k]
  [ 6] perf_ioctl_[k]
  [ 7] do_vfs_ioctl_[k]
  [ 8] ksys_ioctl_[k]
  [ 9] __x64_sys_ioctl_[k]
  [10] do_syscall_64_[k]
  [11] entry_SYSCALL_64_after_hwframe_[k]
  [12] __GI___ioctl

--- 129177 ns (2.86%), 1 sample
  [ 0] _raw_spin_unlock_irqrestore_[k]
  [ 1] pty_write_[k]
  [ 2] n_tty_write_[k]
  [ 3] tty_write_[k]
  [ 4] __vfs_write_[k]
  [ 5] vfs_write_[k]
  [ 6] ksys_write_[k]
  [ 7] __x64_sys_write_[k]
  [ 8] do_syscall_64_[k]
  [ 9] entry_SYSCALL_64_after_hwframe_[k]
  [10] write
  [11] writeBytes
  [12] Java_java_io_FileOutputStream_writeBytes
  [13] java.io.FileOutputStream.writeBytes
  [14] java.io.FileOutputStream.write
  [15] java.io.BufferedOutputStream.flushBuffer
  [16] java.io.BufferedOutputStream.flush
  [17] java.io.PrintStream.write
  [18] sun.nio.cs.StreamEncoder.writeBytes
  [19] sun.nio.cs.StreamEncoder.implFlushBuffer
  [20] sun.nio.cs.StreamEncoder.flushBuffer
  [21] java.io.OutputStreamWriter.flushBuffer
  [22] java.io.PrintStream.write
  [23] java.io.PrintStream.print
  [24] java.io.PrintStream.println
  [25] sun.misc.Version.print
  [26] sun.misc.Version.print

--- 127117 ns (2.81%), 1 sample
  [ 0] native_write_msr_[k]
  [ 1] x2apic_send_IPI_[k]
  [ 2] native_send_call_func_single_ipi_[k]
  [ 3] generic_exec_single_[k]
  [ 4] smp_call_function_single_[k]
  [ 5] task_function_call_[k]
  [ 6] event_function_call_[k]
  [ 7] _perf_event_enable_[k]
  [ 8] _perf_event_refresh_[k]
  [ 9] _perf_ioctl_[k]
  [10] perf_ioctl_[k]
  [11] do_vfs_ioctl_[k]
  [12] ksys_ioctl_[k]
  [13] __x64_sys_ioctl_[k]
  [14] do_syscall_64_[k]
  [15] entry_SYSCALL_64_after_hwframe_[k]
  [16] __GI___ioctl

--- 100018 ns (2.21%), 1 sample
  [ 0] smp_call_function_single_[k]
  [ 1] task_function_call_[k]
  [ 2] perf_install_in_context_[k]
  [ 3] __do_sys_perf_event_open_[k]
  [ 4] __x64_sys_perf_event_open_[k]
  [ 5] do_syscall_64_[k]
  [ 6] entry_SYSCALL_64_after_hwframe_[k]
  [ 7] syscall

--- 98057 ns (2.17%), 1 sample
  [ 0] ktime_get_real_seconds_[k]
  [ 1] __vfs_write_[k]
  [ 2] vfs_write_[k]
  [ 3] ksys_write_[k]
  [ 4] __x64_sys_write_[k]
  [ 5] do_syscall_64_[k]
  [ 6] entry_SYSCALL_64_after_hwframe_[k]
  [ 7] write
  [ 8] writeBytes
  [ 9] Java_java_io_FileOutputStream_writeBytes
  [10] java.io.FileOutputStream.writeBytes
  [11] java.io.FileOutputStream.write
  [12] java.io.BufferedOutputStream.flushBuffer
  [13] java.io.BufferedOutputStream.flush
  [14] java.io.PrintStream.write
  [15] sun.nio.cs.StreamEncoder.writeBytes
  [16] sun.nio.cs.StreamEncoder.implFlushBuffer
  [17] sun.nio.cs.StreamEncoder.flushBuffer
  [18] java.io.OutputStreamWriter.flushBuffer
  [19] java.io.PrintStream.write
  [20] java.io.PrintStream.print
  [21] java.io.PrintStream.println
  [22] sun.misc.Version.print
  [23] sun.misc.Version.print

--- 78800 ns (1.74%), 1 sample
  [ 0] _raw_spin_unlock_irqrestore_[k]
  [ 1] pty_write_[k]
  [ 2] n_tty_write_[k]
  [ 3] tty_write_[k]
  [ 4] __vfs_write_[k]
  [ 5] vfs_write_[k]
  [ 6] ksys_write_[k]
  [ 7] __x64_sys_write_[k]
  [ 8] do_syscall_64_[k]
  [ 9] entry_SYSCALL_64_after_hwframe_[k]
  [10] __write
  [11] [unknown]

--- 68361 ns (1.51%), 1 sample
  [ 0] ThreadShadow::clear_pending_exception()
  [ 1] jni_DetachCurrentThread
  [ 2] JavaMain
  [ 3] start_thread

          ns  percent  samples  top
  ----------  -------  -------  ---
     2195706   48.62%        8  _raw_spin_unlock_irqrestore_[k]
     1040401   23.04%        1  JvmtiEnv::GetClassMethods(oopDesc*, int*, _jmethodID***)
      460051   10.19%        1  ClassFileParser::parse_constant_pool(Thread*)
      395058    8.75%        3  native_write_msr_[k]
      258584    5.73%        3  smp_call_function_single_[k]
       98057    2.17%        1  ktime_get_real_seconds_[k]
       68361    1.51%        1  ThreadShadow::clear_pending_exception()
```

Note that profiler starts only after VM is fully initialized, so it's a very short period of time while profiler is active during execution of `Version.print`.

This question gives me an idea though - what if initialize the profiler earlier at VM startup. It won't be able to sample Java threads, but it could probably collect some info about VM threads. I'll make an experiment to see whether it makes sense.

**Comment by kaiseu** [comment]
@apangin thanks for the information! 
reducing the interval below command works, 
java -agentpath:./build/libasyncProfiler.so=start,summary,flat,traces,interval=100us -version

but when trying to collect some metrics, there still no output file. The command used is: 
for i in cpu lock alloc branch-misses context-switches cache-misses;
do ./profiler.sh -e $i  -d 30 -f $i.svg $pid; done

when using start and then stop, there is some output in the console. 
./profiler.sh start $pid
./profiler.sh stop $pid

**Comment by apangin** [comment]
1. What events are collected, and what events are not?
2. Are there any perf-related warnings or errors in the console/stderr of the target JVM process?
3. Does the above profiler.sh loop work for other Java processes? E.g. what if you try the same approach to profile a simple Java program like [AllocatingTarget.java](https://github.com/jvm-profiling-tools/async-profiler/blob/master/test/AllocatingTarget.java) from the unit tests?

**Comment by kaiseu** [comment]
no events are collected. no warning and errors, no output for other java process.

**Comment by kaiseu** [comment]
find the cause. it seems -f only works with absolute path...

**Comment by apangin** [comment]
Yes, relative paths are relative to the target JVM process.
The problem is understood. I'll close this issue as a duplicate of #229
